File: 02 DATABASES/databases.py
# DATABASES (user,pw are generics)
import pyodbc
import mysql.connector
import psycopg2

# MySQL Con
con_mysql = mysql.connector.connect(
    host= "localhost",
    user= "root",
    password= "root",
    database= "db")

# Postgres con
con_ps = psycopg2.connect(
    host = "localhost",
    port = 5432,
    database = "test",
    user = "postgres",
    password = "test")

# MS SQL con
con_ms = pyodbc.connect(
    "Driver={SQL Server Native Client 11.0};"
    "Server=PC\MSSQL;"
    "Database=testdb;"
     "Trusted_Connection=yes;")

# MS Access con
con_ac = (
    r"DRIVER={Microsoft Access Driver (*.mdb, *.accdb)};"
    r"DBQ=C:\test.accdb;")

# Creating cursor (for mysql, nearly all rdbms work similiar from here on)
cursor = con_mysql.cursor()

# Execute Queries
cursor.execute("USE db")                                                # executes a single SQL query

val = [
    ("Test", "Person")
]

cursor.executemany(""" INSERT INTO users (first_name, last_name)
                       VALUES (%s, %s)""", val)                        # executes multiple query (%s to escape direct input!)

# Read data
cursor.execute("SELECT * FROM customers")               # gets first record from cursor (reset necessary cause fetchall)
print(cursor.fetchone())
cursor.reset()

cursor.execute("SELECT * FROM customers")              # collects n records in array of tuples (size n as argument)
print(cursor.fetchmany(2))
cursor.reset()

cursor.execute("SELECT * FROM customers")              # collects every record in an array of tuples
print(cursor.fetchall())
cursor.reset()

cursor.execute("SELECT * FROM customers")              # iterates over every record in the cursor
for row in cursor:
    print(row)
cursor.reset()

# Commit (otherwise everything above won't be written into Database)
con_mysql.commit()

File: 02 DATABASES/databases.js
// MySQL (npm install mysql)
var mysql = require("mysql");

var con = mysql.createConnection({
    host: "localhost",
    user: "root",
    password: "root",
    database: "db"
})

con.connect()

con.query("SELECT * FROM users", (err, res, fields) => {        // query takes SQL Query as String and Callback function as arg (here arrow)
    res.forEach(record => {                                     // mysql returns an array of objects
        console.log(record, record["id"], record.id)            // specific columns can be accessed with []
    })
})

con.commit()                                                    // saving all changes
con.end()                                                       // end connection(!)


// Postgres (npm install pg)
var pg = require("pg")

var con_ps = new pg.Pool({
    host: "localhost",
    user: "postgres",
    password: "password",
    port: 5432,
    database: "test"
})

con_ps.query("SELECT * FROM mytest", (err, res) => {
    res["rows"].forEach(record => {                     // Postgres returns an object. The records are stored in the "rows" attribute as an array of objects
        console.log(record, record["id"], record.id)    // specific columns can be accessed with []
    })
})

con_ps.end()                                           // end connection(!)


// MS SQL (npm install mssql + npm install msnodesqlv8)
var ms = require("mssql/msnodesqlv8");

var con_ms = new ms.ConnectionPool({
    database: "testdb",
    server: "localhost",
    options: {
        trustedConnection: true                                     // using windows auth here
    }
})

con_ms.connect(() => {                                              // haven't found a way to 'hold' the connection open like in mysql/ps
    con_ms.request().query("SELECT * FROM test", (err, res) => {
        console.log(res["recordset"])                               // returns an object where records are stored in recordset as an array of objects
        console.log(res["recordset"][0]["name"])                    // specific columns can be accessed with []

        con_ms.close()                                              // end connection(!)
    })
})
